src/contract_response.ts

```diff
@@ -0,0 +1,186 @@
+    throw new Error(this.error.message);
+  }
+
+  isOk(): boolean {
+    return false;
+  }
+
+  isErr(): boolean {
+    return !this.isOk();
+  }
+}
+export class Resources {
+  fee: number;
+  refundableFee: number;
+  cpuInst: number;
+  readBytes: number;
+  writeBytes: number;
+  readOnlyEntries: number;
+  readWriteEntries: number;
+
+  constructor(
+    fee: number,
+    refundableFee: number,
+    cpuInst: number,
+    readBytes: number,
+    writeBytes: number,
+    readOnlyEntries: number,
+    readWriteEntries: number
+  ) {
+    this.fee = fee;
+    this.refundableFee = refundableFee;
+    this.cpuInst = cpuInst;
+    this.readBytes = readBytes;
+    this.writeBytes = writeBytes;
+    this.readOnlyEntries = readOnlyEntries;
+    this.readWriteEntries = readWriteEntries;
+  }
+
+  /**
+   * Builds a Resources object from TransactionEnvelope
+   * @returns - Resources
+   */
+  static fromTransaction(tx: xdr.TransactionEnvelope | string): Resources {
+    if (typeof tx === 'string') {
+      tx = xdr.TransactionEnvelope.fromXDR(tx, 'base64');
+    }
+    const transaction = tx.v1().tx();
+    const data = transaction.ext().sorobanData();
+    const sorobanResources = data.resources();
+    const footprint = sorobanResources.footprint();
+
+    const fee = transaction.fee();
+    const refundableFee = Number(data.resourceFee().toString());
+    const cpuInst = sorobanResources.instructions();
+    const readBytes = sorobanResources.readBytes();
+    const writeBytes = sorobanResources.writeBytes();
+    const readOnlyEntries = footprint.readOnly().length;
+    const readWriteEntries = footprint.readWrite().length;
+
+    return new Resources(
+      fee,
+      refundableFee,
+      cpuInst,
+      readBytes,
+      writeBytes,
+      readOnlyEntries,
+      readWriteEntries
+    );
+  }
+}
+
+export class ContractResponse<T> {
+  result: Result<T, ContractError>;
+  hash: string;
+  resources: Resources;
+
+  private constructor(result?: Result<T, ContractError>) {
+    if (result) {
+      this.result = result;
+    }
+  }
+  static fromSimulationResponse<T>(
+    simulation: SorobanRpc.Api.SimulateTransactionResponse,
+    transaction: Transaction<Memo<MemoType>, Operation[]> | string,
+    network_passphrase: string,
+    parser: (xdr: string) => T
+  ): ContractResponse<T> {
+    let response = new ContractResponse<T>();
+    if (typeof transaction === 'string') {
+      transaction = new Transaction(transaction, network_passphrase);
+    }
+    response.hash = transaction.hash().toString('hex');
+
+    if (SorobanRpc.Api.isSimulationError(simulation)) {
+      response.result = new Err(parseError(simulation));
+      response.resources = new Resources(0, 0, 0, 0, 0, 0, 0);
+    } else if (SorobanRpc.Api.isSimulationRestore(simulation)) {
+      response.result = new Err(
+        new ContractError(
+          ContractErrorType.InvokeHostFunctionEntryArchived,
+          JSON.stringify(simulation.restorePreamble)
+        )
```

Comment:
this might not be relevant for this PR, but we might want to keep this object in tact?

It could make sense to have an optional `restorePreamble` field or something we can append this too, and verify we can actually create the correct transaction from it.

Reply:
The consumer should still have the simulation response so they would be able to use it to build the restore transaction.

Reply:
Good point, forgot about that. In this case - we should put a more reasonable error message in here.

(the json string is likely massive given its a bunch of XDR types).

Consider `invokeHostFunctionEntryArchived` or something?